### PR TITLE
Issue #634 : Retrieve and display domains of ConceptScheme at the top of the tree hierarchy

### PR DESCRIPTION
**Pull request by tfrancart**
Coming back with a PR for #634 .
I have made this cleaner.
I am resubmitting the PR, hoping it can be useful and integrated in the code.
This answers 2 client use-cases I have (UNESCO and another).

Could you please let me know how can this proposal can progress to be integrated in the codebase ?

### CODE REVIEW

File: controller/RestController.php [hunk 1]
```diff
@@ -200,6 +200,7 @@ public function vocabularyInformation($request)
                 'skos' => 'http://www.w3.org/2004/02/skos/core#',
                 'onki' => 'http://schema.onki.fi/onki#',
                 'dct' => 'http://purl.org/dc/terms/',
+                'dcterms' =>'http://purl.org/dc/terms/',
```
Comment (osma) [inline controller/RestController.php]: Why do you need to define this prefix? `dct` is defined on the line above, you can just use that.

File: controller/RestController.php [hunk 2]
```diff
@@ -208,6 +209,7 @@ public function vocabularyInformation($request)
                 'defaultLanguage' => 'onki:defaultLanguage',
                 'languages' => 'onki:language',
                 'label' => 'rdfs:label',
+                'subject' => 'dcterms:subject',
```
Comment (osma) [inline controller/RestController.php]: Use dct

File: controller/RestController.php [hunk 3]
```diff
@@ -254,6 +256,7 @@ public function vocabularyStatistics($request)
                 'void' => 'http://rdfs.org/ns/void#',
                 'onki' => 'http://schema.onki.fi/onki#',
                 'uri' => '@id',
+                'dcterms' =>'http://purl.org/dc/terms/',
```
Comment (osma) [inline controller/RestController.php]: Ditto

File: controller/RestController.php [hunk 4]
```diff
@@ -549,6 +552,7 @@ private function returnDataResults($results, $format) {
                 'dct' => 'http://purl.org/dc/terms/',
                 'dc11' => 'http://purl.org/dc/elements/1.1/',
                 'uri' => '@id',
+                'dcterms' =>'http://purl.org/dc/terms/',
```
Comment (osma) [inline controller/RestController.php]: Ditto

File: model/GlobalConfig.php [hunk 5]
```diff
@@ -1,243 +1,243 @@
-<?php
```
Comment (osma) [inline model/GlobalConfig.php]: Please don't make style fixes in unrelated code as part of a feature PR. It's practically impossible to see whether you've changed anything in this file, making it difficult to review. I suggest you revert this part and make it a separate PR if you like.

File: model/Model.php [hunk 6]
```diff
@@ -57,7 +57,6 @@ public function __construct($config)
     public function getConfig() {
       return $this->globalConfig;
     }
-
```
Comment (osma) [inline model/Model.php]: Same as above, even though this is rather harmless...

File: model/Vocabulary.php [hunk 7]
```diff
@@ -229,6 +229,21 @@ public function getDefaultConceptScheme()
         return array_pop($conceptSchemeURIs);
     }
 
+    /**
+     * Return the URI of the default concept scheme of this vocabulary, as set in the vocabulary configuration, or null
+     * if not set.
+     * @return string default concept scheme URI, or null if undefined
+     **/
+
+    public function getMainConceptScheme(){
```
Comment (osma) [inline model/Vocabulary.php]: This is partially duplicating `getDefaultConceptScheme`. Could you also make `getDefaultConceptScheme` take advantage of this function to avoid the duplication. (i.e. if getMainConceptScheme() returns null, guess the concept scheme)

File: model/sparql/GenericSparql.php [hunk 8]
```diff
@@ -31,6 +31,9 @@ class GenericSparql {
      */
     private $qnamecache = array();
 
+
```
Comment (osma) [inline model/sparql/GenericSparql.php]: Unnecessary whitespace

File: model/sparql/GenericSparql.php [hunk 9]
```diff
@@ -595,9 +598,14 @@ public function queryConceptScheme($conceptscheme) {
     private function generateQueryConceptSchemesQuery($lang) {
         $fcl = $this->generateFromClause();
         $query = <<<EOQ
-SELECT ?cs ?label ?preflabel ?title $fcl
+SELECT ?cs ?label ?preflabel ?title ?domain ?domainLabel $fcl
 WHERE {
  ?cs a skos:ConceptScheme .
+ OPTIONAL{
+    ?cs dcterms:subject ?domain.
```
Comment (osma) [inline model/sparql/GenericSparql.php]: You should be able to use `dct:subject` here

File: resource/js/hierarchy.js [hunk 10]
```diff
@@ -128,76 +128,126 @@ function createConceptObject(conceptUri, conceptData) {
  * @param {Object} parentData 
  */
 function attachTopConceptsToSchemes(schemes, currentNode, parentData) {
```
Comment (osma) [inline resource/js/hierarchy.js]: The changes in this file look more invasive than they are, because you've reformatted the code when editing it. Consider making it more like the original code so it's easier to see what has changed. However, I understand that this may be difficult to do afterwards.

File: resource/js/hierarchy.js [hunk 11]
```diff
@@ -294,20 +347,96 @@ function pickLabelFromScheme(scheme) {
...
+        // Step 2.2 : find the concept schemes in this domain and add them as children
+        for (var k = 0; k < schemes.length; k++) {
+        	var theScheme = schemes[k];
+        	var theSchemeLabel = pickLabel(theScheme);
+        	
+        	// avoid creating entries with empty labels
+        	if(theSchemeLabel != '') { 
+	        	if((theScheme.subject) != null && (theScheme.subject.uri===theDomain.uri)) {
+	        		domainObject.children.push(
+	        				{
```
Comment (osma) [inline resource/js/hierarchy.js]: The indents look rather excessive here.

File: resource/js/hierarchy.js [hunk 12]
```diff
@@ -427,7 +557,29 @@ function getTreeConfiguration() {
...
+    			else if(nodeA.a_attr['class'] != 'domain' && nodeB.a_attr['class'] == 'domain') {
+    				return 1;
+    			}
+    		} else {
+    			return naturalCompare(this.get_text(a).toLowerCase(), this.get_text(b).toLowerCase());
+    		}
+    	}  else {
+    		return naturalCompare(this.get_text(a).toLowerCase(), this.get_text(b).toLowerCase());
+    	}
+    	// return naturalCompare(this.get_text(a).toLowerCase(), this.get_text(b).toLowerCase());
```
Comment (osma) [inline resource/js/hierarchy.js]: Please don't leave old code as comments. We have versioning for that.

File: tests/GenericSparqlTest.php [hunk 13]
```diff
@@ -350,6 +350,49 @@ public function testQueryConceptSchemes()
       $this->assertEquals('Test conceptscheme', $label['label']);
     }
   }
+
```
Comment (osma) [inline tests/GenericSparqlTest.php]: It's excellent that you've written tests for your new functionality!

File: tests/GenericSparqlTest.php [hunk 14]
```diff
@@ -738,7 +781,7 @@ public function testQueryParentList()
     );
     $props = array (
       'uri' => 'http://www.skosmos.skos/test/ta1',
-      'top' => 'http://www.skosmos.skos/test/conceptscheme',
```
Comment (osma) [inline tests/GenericSparqlTest.php]: Please don't drop the definition of `top`, it is still needed for backwards compatibility.

**Comment by osma** [comment]
Also, please rebase on top of current master. There are conflicts so this cannot be merged as-is.
Travis CI build failed, please take a look at that. I suspect the GlobalConfig changes might be the reason but not sure.

**Comment by tfrancart** [comment]
- sigh - 
You're probably right, but "rebasing" or "squashing" or modifying files without adding yet-another commit to the commit history is just something I don't know how to do properly.
I will close this PR, manually create another branch from master, re-do the modifications in a cleaner way in the new code, break it down in smaller chunks as you suggest, and open a new PR.

**Comment by osma** [comment]
Thanks for your patience!
Squashing is optional, it will just make the commit history cleaner by merging multiple successive commits into one. It's very easy to do (git rebase -i) and it's literally the opposite of "adding yet-another commit" :)
But the important part is not whether the history looks clean (it can be fixed later if necessary), it's that the PR applies cleanly (no conflicts), passes Travis checks and is easy enough to review. Contributing small bits at a time helps all these goals.
